src/vs/editor/common/commands/replaceCommand.ts

```diff
@@ -31,6 +32,73 @@ export class ReplaceCommand implements ICommand {
 	}
 }
 
+export class ReplaceOvertypeCommand implements ICommand {
+
+	private readonly _range: Range;
+	private readonly _text: string;
+	public readonly insertsAutoWhitespace: boolean;
+
+	constructor(range: Range, text: string, insertsAutoWhitespace: boolean = false) {
+		this._range = range;
+		this._text = text;
+		this.insertsAutoWhitespace = insertsAutoWhitespace;
+	}
+
+	public getEditOperations(model: ITextModel, builder: IEditOperationBuilder): void {
+		const startPosition = this._range.getStartPosition();
+		const endPosition = this._range.getEndPosition();
+		const rangeEndOffset = model.getOffsetAt(endPosition);
+		const endOffset = rangeEndOffset + this._text.length + (this._range.isEmpty() ? 0 : - 1);
+		const endOfLine = model.getEndOfLineSequence() === EndOfLineSequence.CRLF ? '\r\n' : '\n';
+		const lastCharacterRange = Range.fromPositions(model.getPositionAt(endOffset - 1), model.getPositionAt(endOffset));
+		const lastCharacter = model.getValueInRange(lastCharacterRange);
+		const newEndOffset = lastCharacter === endOfLine ? endOffset - 1 : endOffset;
+		const replaceRange = Range.fromPositions(startPosition, model.getPositionAt(newEndOffset));
+		builder.addTrackedEditOperation(replaceRange, this._text);
+	}
```

Comment:
In principle, we should avoid using offsets for simple command implementations because converting coordinates from (line,col) to offsets has some cost (although logarithmic, there is still some cost), and when doing something like 1000 multi-cursors and typing a keystroke, this cost multiplies. Is it possible to express this logic without converting to offsets?